### ISSUE TITLE
CONNECTIONIST AI - Basics of Neural Networks (Cornelius Jabez Lim)

### ISSUE THREAD

**Issue by timothypardin28** [BIT-s-Absolute-Beginners/Absolute-Beginners---AI-CS-BIG-DATA-CONNECTIONIST-AI-#25]

**Comment by corneljz** [comment]
# **Chapter 0 | Introduction to Neural Network (NN)**
> Reference: [Neural Network In 5 Minutes | What Is A Neural Network? | How Neural Networks Work | Simplilearn](https://www.youtube.com/watch?v=bfmFfD2RIcg)

1. The Base of Deep Learning, a subfield of machine learning.
2. An algorithm based on the structure of the human brain

Take Data -> Train with the data -> Predict Data Output

## Components
1. Neurons: NN is made of layers of neurons. It is the core processing unit of the network.
2. The Input Layer (The First Layer): Receives the input / data.
3. The Hidden Layer (The Ones in the middle): Perform most of the computation required by the network.
4. The Output Layer (The Last Layer): Predicts the final output.

## How it Works (Example: Determine shapes)!
1. A circle image is broken down into pixels.
2. Each pixel value is attached to a node/neuron in the input layer.
3. Neurons of the first layer are connected to the next layer (hidden layer) through channels
4. Each channel is assigned a numerical value known as weight The inputs are multiply to the coresponding weights.
5. Their sum is sent as input to the hidden layer neurons
6. Each of the neurons is associated to a numerical value called the bias
7. The bias is added to the input sum
8. The value of the equation is then pass into a threshold function called the activation function. The result will determine whether the particular neuron will be activated or not.
9. An activated neuron will transmit its data to the neuron of the next layer though the channels 
10. The data is propagated (transmitted) through the network
11. This is called forward propagation
12. In the out put layer, the neuron with the highest value (probability) fires and determines the output.
13. The network would most probably made the wrong decision/answer
14. The output is compared to the actual output to realise the error in prediction.
15. The magnitude of the error indicates how wrong it is, usually using the Loss Function to determine how far the answer is from the real one.
16. The gathered information is sent backwards (Backpropagation)
17. Based on the information the wights are adjusted 
18. This cycle continues on with multiple inputs until the answer is correct or satisfying.

## Prime Application of NN
1. Facial Recognition
2. Forecasting
3. Music Composition

---

---

---

> Reference: [Neural Networks | 3Blue1Brown](https://www.youtube.com/watch?v=aircAruvnKk&list=PLZHQObOWTQDNU6R1_67000Dx_ZCJB-3pi)

# **CHAPTER 1: What is Neural Network?**
1. Inspired by the brain

## What Are Neural Networks?
### Components
1. Neurons: A function that takes in the previous neuron or data and spit-out a value between 0 and 1
2. Weights: A value that indicates the connection between the current neuron and the previous neurons. Each connection can have different weights to signify its neuron-to-neuron connectivity.
3. Network: A function that takes in data and outputs a result accordingly.

## Learning the Structure
### The First Layer (Input Layer)
1. Input a picture of a number (ex: 28x28 pixel).
2. Each neurons correspond to one pixel in the picture (so 784 neurons).
3. The neurons holds a value that represent the greyscale value of its corresponding pixels; 0 for black and 1 for white.
4. The value that the neurons hold is called activation. So, when the activation is a high number, the neuron will "light up".
5. These receiving neurons is the first layer of our network.

### The Last Layer (Output Layer)
1. Now, the last layer has the same number of the option that we have. In this case 10 for ten number from 0-9.
2. In this case, the value of these neurons correspond to how the system thinks that the given image corresponds with a given digit.

### The Hidden Layer (The Layers in Between)
1. Every neuron here is connected with all the neuronsof the previous layer.
2. Each have a bias number.

### GENERAL Structure Knowledge
- The activation of the previous layer determines the activation of the next layer.
- Basically the first pattern cause the next pattern to make a specific pattern that at the end have a final pattern that correspond, hopefully, to the right output.
- Here neurons are assigned with different weights to signify its relation with a new neuron in a different layer.
- To know the total or the sum of weights, each neuron value must be multiplied by its corresponding weight. The total weight is the sum of them.
- The total weight could be any number, so a solution is to some how squish its value to a range of 0 and 1. A way to do this is through using a sigmoid function.
- So, the hidden layer neuron value is to measure how positive the relevant weighted sum is.
-To put a limit, we can simply subtract the value of the sum weights with a specific number called bias before inserting it into the sigmoid function.
#### One Possible Logic Behind
How do know how much hidden layers and neuron we should have?
1. Understand the possible logic. To produce a number, we can breakdown a number into several pieces. Let say a cricle, a straight line, and so on.
2. These pieces can be further broken down into smaller pieces. A circle can be broken down into several edges.
3. The smaller edges can be the value of the first hidden layer, and the bigger pieces is the second one.
4. When the input receives an image, the neuron will receive its pixels greyscale value. The specific pattern that light up can create edges recognise by the first hidden layer.
5. In this hidden layer, the neurons corresponding to the specific pattern will light up. This value is then tranferred to the next layer which will have the possible combination of the smaller edges into bigger pieces.
6. These bigger pieces is key because to the output with the biggest probability is the one that matches the specification based on the combination from the previous layer.
7. Figuring the combination of pieces can lead to a higher probability of one output.

### No Longer Sigmoid Function in Modern Life, Hello ReLU!
ReLU : Rectified Linear Unit
ReLU takes the max of 0 and a, so equal or below zero, the function will output 0. While above zero, the function will output a number that signifies that the value of the neuron is 1 or light up.
![Sigmoid vs ReLU](Sigmoid%20vs%20ReLU.png)

### What is Learning in Neural Networks?
1. Letting the computer find the right weights and biases.
**THIS SECTION WILL BE EXPLAIN IN THE NEXT CHAPTER BELOW**

### The Grand Picture of What is Happening in the Hidden Layer Neurons:
![Matrix Sigmoid Hidden Layer Neuron](Matrix%20and%20Sigmoid.png)
![Matrix Sigmoid Hidden Layer Neuron 2](Matrix%20and%20Sigmoid%202.png)

# **Chapter 2: How Neural Networks Learn**
## Neural Network Function
### Simple Processing Explanation
Input: Data In
Output: Possible Answer
Parameters: Number of weights & biases
## Cost Function
### Simple Processing Explanation
Input: Number of weights & biases
Output: 1 Number (the cost)
Parameters: Many training example
### A Brief
1. Compare the answer given by the network with the real answer.
2. Add-up the squares of the differences between the network and the real answer (for each neuron).
3. The sum is small when the system is near the real answer
4. Find the average cost of all training data as a reference also for the computer to know how bad it is.

## Telling The Computer How to Change the Weights & Biases
